template_functions.go

```diff
@@ -231,6 +231,43 @@ func treeFunc(brain *Brain,
 
 }
 
+// byKey accepts a slice of KV pairs and returns a map of the top-level
+// key to all its subkeys. For example:
+//
+//		elasticsearch/a //=> "1"
+//		elasticsearch/b //=> "2"
+//		redis/a/b //=> "3"
+//
+// Passing the result from Consul through byTag would yield:
+//
+// 		map[string]map[string]string{
+//	  	"elasticsearch": &dep.KeyPair{"a": "1"}, &dep.KeyPair{"b": "2"},
+//			"redis": &dep.KeyPair{"a/b": "3"}
+//		}
+//
+// Note that the top-most key is stripped from the Key value. Keys that have no
+// prefix after stripping are removed from the list.
+func byKey(pairs []*dep.KeyPair) (map[string][]*dep.KeyPair, error) {
+	m := make(map[string][]*dep.KeyPair)
+	for _, pair := range pairs {
+		parts := strings.Split(pair.Key, "/")
```

Comment:
I originally tried this, but the API promise for `SplitN` is "at most", not "at least", so it's possible for the result to easily be out of range (if given something that doesn't have a slash for example). As a result, I had to hack together this:

```go
		var top, key string
		if len(parts) < 2 {
			top = parts[0]
			key = ""
		} else {
			top = parts[0]
			key = parts[1]
		}
```

which just felt really wrong to me. If there's a way to guarantee that `SplitN` returns that number of elements, that would be great, but it doesn't look like it can.